### PR TITLE
New ask-or-tell.

### PR DESCRIPTION
**Pull request by anastasiamac**
## Description of change

This PR changes cloud level commands (CRUD for clouds, credentials and k8s). 
Since these commands can operate on both client and controller copies of the entities, the ambiguity of what user intends needs to be clarified. Thus, when user supplies:
* ***default, i.e. no options supplied***: User will be prompted to confirm whether the operation is to take place on the client and or controller (and if on controller, then which one). Once the answers are obtained, the operation will proceed accordingly;
* ***--client***: the operation will take place on the client;
* ***-c/--controller <controller name>***: the operation will take place on the controller.

This approach also allows the combination of the options, i.e. when both --client and -c options are correctly provided, the operation will take place on both the client and the controller.

For this reason, --client-only option has been renamed to --client.

## Bug reference

https://bugs.launchpad.net/bugs/1849601

### CODE REVIEW

**Comment by anastasiamac** [comment]
[new ask-or-tell](https://trello.com/c/ETLjraHK/963-new-ask-or-tell)

File: cmd/juju/cloud/listcredentials.go [hunk 1]
```diff
@@ -243,15 +235,22 @@ func (c *listCredentialsCommand) Run(ctxt *cmd.Context) error {
 		ctxt.Infof("secrets are not shown in tabular format")
 		c.showSecrets = false
 	}
-	credentials := credentialsMap{ClientOnly: c.ClientOnly}
+	credentials := credentialsMap{ClientOnly: c.Client}
+	cloudMsg := ""
+	if c.cloudName != "" {
+		cloudMsg = fmt.Sprintf("for cloud %q ", c.cloudName)
+	}
+	if err := c.MaybePrompt(ctxt, fmt.Sprintf("list credentials %vfrom", cloudMsg)); err != nil {
```
Comment (wallyworld) [inline cmd/juju/cloud/listcredentials.go]: this prompt will be wrong if c.cloudName == "" since cloudMsg will be ""

File: cmd/juju/cloud/listcredentials.go [hunk 2]
```diff
@@ -383,11 +366,12 @@ func formatCredentialsTabular(writer io.Writer, value interface{}) error {
 		return errors.Errorf("expected value of type %T, got %T", credentials, value)
 	}
 
-	if len(credentials.Client) == 0 {
-		fmt.Fprintln(writer, "No credentials from this client to display.")
-	}
-	if len(credentials.Controller) == 0 {
-		if !credentials.ClientOnly {
+	if credentials.ClientOnly {
```
Comment (wallyworld) [inline cmd/juju/cloud/listcredentials.go]: I'm unclear why the change from credentials.Client to credentials.ClientOnly

File: cmd/juju/cloud/remove.go [hunk 3]
```diff
@@ -102,33 +93,23 @@ func (c *removeCloudCommand) Init(args []string) (err error) {
...
-			if err != nil {
-				return errors.Trace(err)
-			}
-		}
+	if err := c.MaybePrompt(ctxt, fmt.Sprintf("remove cloud %v from", c.Cloud)); err != nil {
+		return errors.Trace(err)
 	}
-	if c.ControllerName == "" && !c.ClientOnly {
-		ctxt.Infof("To remove cloud %q from this client, use the --client-only option.", c.Cloud)
+	if c.ControllerName == "" && !c.Client {
```
Comment (wallyworld) [inline cmd/juju/cloud/remove.go]: Here is a case where I think the logic is correct now but was wrong originally?

File: cmd/juju/cloud/remove.go [hunk 4]
```diff
@@ -102,33 +93,23 @@ func (c *removeCloudCommand) Init(args []string) (err error) {
...
+		ctxt.Infof("To remove cloud %q from this client, use the --client option.", c.Cloud)
```
Comment (wallyworld) [inline cmd/juju/cloud/remove.go]: We can return here right?
If not controller and not client then nothing to do.
Should this logic be used elsewhere? It seems this is the only place where we do this check and can exit early? Maybe I'm wrong, lots of code to look at

File: cmd/juju/cloud/remove_test.go [hunk 5]
```diff
@@ -144,8 +140,9 @@ func (s *removeSuite) TestRemoveCloudController(c *gc.C) {
 
 func (s *removeSuite) TestCannotRemovePublicCloud(c *gc.C) {
 	s.createTestCloudData(c)
-	ctx, err := cmdtesting.RunCommand(c, cloud.NewRemoveCloudCommand(), "prodstack", "--client-only")
-	c.Assert(err, jc.ErrorIsNil)
+	ctx, _ := cmdtesting.RunCommand(c, cloud.NewRemoveCloudCommand(), "prodstack", "--client")
+	//ctx, err := cmdtesting.RunCommand(c, cloud.NewRemoveCloudCommand(), "prodstack", "--client")
```
Comment (wallyworld) [inline cmd/juju/cloud/remove_test.go]: remove me

File: cmd/juju/cloud/removecredential.go [hunk 6]
```diff
@@ -127,22 +116,14 @@ func (c *removeCredentialCommand) SetFlags(f *gnuflag.FlagSet) {
...
+		return errors.Trace(err)
 	}
-	if c.ControllerName == "" && !c.ClientOnly {
-		ctxt.Infof("To remove credential %q for cloud %q from this client, use the --client-only option.", c.credential, c.cloud)
+	if c.ControllerName == "" && !c.Client {
+		ctxt.Infof("To remove credential %q for cloud %q from this client, use the --client option.", c.credential, c.cloud)
 	}
 	var client RemoveCredentialAPI
-	if !c.ClientOnly {
+	if c.ControllerName != "" {
```
Comment (wallyworld) [inline cmd/juju/cloud/removecredential.go]: This logic can be moved down to the other if controllerName != "" block

File: cmd/juju/cloud/show.go [hunk 7]
```diff
@@ -176,7 +160,7 @@ func (c *showCloudCommand) Run(ctxt *cmd.Context) error {
 			}
 		}
 	}
-	if c.BothClientAndController || c.ClientOnly {
+	if c.Client {
```
Comment (wallyworld) [inline cmd/juju/cloud/show.go]: if c.Client && localCloud != nil
Actually, just if localCloud != nil is sufficient right?
And same with remoteCloud != nil above

File: cmd/juju/cloud/updatecloud.go [hunk 8]
```diff
@@ -145,20 +138,11 @@ func (c *updateCloudCommand) Run(ctxt *cmd.Context) error {
...
-				return errors.Trace(err)
-			}
-		}
+	if err := c.MaybePrompt(ctxt, fmt.Sprintf("update cloud %q on", c.Cloud)); err != nil {
+		return errors.Trace(err)
 	}
-	if c.ControllerName == "" && !c.ClientOnly {
-		ctxt.Infof("To update cloud %q on this client, use the --client-only option.", c.Cloud)
+	if c.ControllerName == "" && !c.Client {
+		ctxt.Infof("To update cloud %q on this client, use the --client option.", c.Cloud)
```
Comment (wallyworld) [inline cmd/juju/cloud/updatecloud.go]: we can return early here etc

File: cmd/juju/cloud/updatecredential.go [hunk 9]
```diff
@@ -35,16 +35,9 @@ a model was created with to the new and valid details on controller.
...
-If the prompt is not needed and the cloud is always to be updated on
-the current controller if that controller is detected, use --no-prompt option.
+Use --controller option to update a cloud on a controller. 
 
-Use --controller option to update a cloud on a different controller. 
-
-Use --controller-only option to only update controller copy of the cloud.
-
-If --client-only is used, Juju updates credential only on this client.
+If --client is used, Juju updates credential on this client.
```
Comment (wallyworld) [inline cmd/juju/cloud/updatecredential.go]: This wording differs from elsewhere

File: cmd/modelcmd/controller.go [hunk 10]
```diff
@@ -418,8 +407,7 @@ type OptionalControllerCommand struct {
 // SetFlags initializes the flags supported by the command.
 func (c *OptionalControllerCommand) SetFlags(f *gnuflag.FlagSet) {
 	c.CommandBase.SetFlags(f)
-	f.BoolVar(&c.ClientOnly, "client-only", false, "Client operation only; controller not affected")
-	f.BoolVar(&c.ControllerOnly, "controller-only", false, "Controller operation only; client not affected")
+	f.BoolVar(&c.Client, "client", false, "Command to operate on or with a copy from this client")
```
Comment (wallyworld) [inline cmd/modelcmd/controller.go]: i don't parse the description

File: cmd/modelcmd/controller.go [hunk 11]
```diff
@@ -429,74 +417,81 @@ func (c *OptionalControllerCommand) SetFlags(f *gnuflag.FlagSet) {
 
 // Init populates the command with the args from the command line.
 func (c *OptionalControllerCommand) Init(args []string) (err error) {
-	if c.Local && !c.ClientOnly {
-		c.ClientOnly = c.Local
+	if c.Local && !c.Client {
```
Comment (wallyworld) [inline cmd/modelcmd/controller.go]: c.Client = c.Client || c.Local

File: cmd/modelcmd/controller.go [hunk 12]
```diff
@@ -429,74 +417,81 @@ func (c *OptionalControllerCommand) SetFlags(f *gnuflag.FlagSet) {
...
-	use, err := pollster.YN(fmt.Sprintf("Do you want to %v current controller %q", action, controllerName), true)
-	if err != nil {
-		return "", errors.Trace(err)
+	if defaultController == "" {
+		defaultController = controllers[0]
 	}
-	if use {
-		return controllerName, nil
+	cName, err := pollster.SelectVerify(interact.List{
+		Singular: "controller name",
```
Comment (wallyworld) [inline cmd/modelcmd/controller.go]: Just "controller" and "controllers"

File: cmd/juju/cloud/listcredentials.go (hunk 1 above)
Comment (anastasiamac) [reply cmd/juju/cloud/listcredentials.go]: Yes, it is intentional and reads fine.
If cloud name is given it'll read 
```
list credentials for cloud AAAA from [client/controller]
````
If not, it'll read
```
list credentials from [cloud/controller]
```

I've experimented with this quiet a bit to get the wording and spacing correct.

File: cmd/juju/caas/remove.go [hunk 13]
```diff
@@ -25,21 +25,11 @@ Removes the specified k8s cloud from this client.
...
-a user will be prompted to confirm if the k8s cloud needs to be removed from it. 
-If the prompt is not needed and the k8s cloud is always to be removed from
-the current controller if that controller is detected, use --no-prompt option.
-
-If you just want to update your controller and not
-your current client, use the --controller-only option.
-
-If you just want to update your current client and not
-a running controller, use the --client-only option.
+Use --client option to update your current client.
```
[the remark this answers is not in the document]
Comment (anastasiamac) [reply cmd/juju/caas/remove.go]: In all our commands that I have just looked at, we do not add an article in this cases.

I have done the search on "Use -" and we apply phrase it this way everywhere - not just cloud and caas package but applications, backups, action, etc.

File: cmd/juju/caas/remove.go [hunk 14]
```diff
@@ -25,21 +25,11 @@ Removes the specified k8s cloud from this client.
...
 
 Examples:
     juju remove-k8s myk8scloud
-    juju remove-k8s myk8scloud --client-only
-    juju remove-k8s myk8scloud --controller-only --no-prompt
+    juju remove-k8s myk8scloud --client
```
[the remark this answers is not in the document]
Comment (anastasiamac) [reply cmd/juju/caas/remove.go]: I do not think so... We do not always have all possible combinations and permutations of options.

File: cmd/juju/caas/remove_test.go [hunk 15]
```diff
@@ -124,63 +122,31 @@ func (s *removeCAASSuite) TestRemove(c *gc.C) {
...
 
 	// client side operations
-	s.cloudMetadataStore.CheckNoCalls(c)
-	s.store.CheckNoCalls(c)
+	s.cloudMetadataStore.CheckCallNames(c, "PersonalCloudMetadata")
+	s.store.CheckCallNames(c, "CredentialForCloud")
 }
 
-func (s *removeCAASSuite) TestRemoveLocalOnly(c *gc.C) {
+func (s *removeCAASSuite) TestRemoveClientOnly(c *gc.C) {
```
[the remark this answers is not in the document]
Comment (anastasiamac) [reply cmd/juju/caas/remove_test.go]: This test checks that the change is only on the client so the naming still makes sense.

File: cmd/juju/cloud/listcredentials.go (hunk 2 above)
Comment (anastasiamac) [reply cmd/juju/cloud/listcredentials.go]: Here .Client has the list of credentials whereas .ClientOnly holds whether the client list was requested or not. The difference is in principal and is needed in the formatter when we get an empty list - is it that there are no credentials on the client or is it that we were not even looking on the client for credentials.

Comment (anastasiamac) [reply cmd/juju/cloud/listcredentials.go]: I've added comments to clarify.

File: cmd/juju/cloud/remove.go (hunk 4 above)
Comment (anastasiamac) [reply cmd/juju/cloud/remove.go]: Actually, I do not think that we need it anymore...

It looks like long ago, the original intent of this code was to detect that the controller and client operation was requested but no controller was specified/detected. In this case, in the next run of this command, the user could have just said "--client"... This is what this message is about - the information/education of the user. 

I do not think it's needed anymore as we no longer detect controllers and users need to specify them explicitly.

File: cmd/juju/cloud/removecredential.go (hunk 6 above)
Comment (anastasiamac) [reply cmd/juju/cloud/removecredential.go]: No, it can't since the line right after it uses this client. We are checking remote clouds and needed to open that connection.

File: cmd/juju/cloud/show.go (hunk 7 above)
Comment (anastasiamac) [reply cmd/juju/cloud/show.go]: Well, no. I've simplified the section a little.

But basically we need to know whether user has requested a client cloud or not, especially in case where we just did not find it.

This is needed because we are displaying config after the cloud details. If both controller and client clouds were requested and since we display controller clouds first, we need to know if we are showing the config of remote cloud or not.

File: cmd/juju/cloud/updatecloud.go (hunk 8 above)
Comment (anastasiamac) [reply cmd/juju/cloud/updatecloud.go]: No, this was information message. And anyway, it's gone now.

File: cmd/juju/cloud/updatecredential.go (hunk 9 above)
Comment (anastasiamac) [reply cmd/juju/cloud/updatecredential.go]: yes, some original wording makes command help unique. The same is on k8s commands. I've updated what I needed here.

File: cmd/modelcmd/controller.go (hunk 10 above)
Comment (anastasiamac) [reply cmd/modelcmd/controller.go]: Changed to "Client operation" echoing deprecated "Local operation only,..." and "Controller to operate in"

**Comment by anastasiamac** [comment]
$$merge$$
